File: src/gmail/subagent.ts
import type { FakeGmail } from './fake.js';

/**
 * Free-form instruction sent to the Gmail sub-agent.
 * The executor builds this from the approved goal.
 */
export interface SubAgentInstruction {
  capability: string;
  messageId: string;
  /** Free-form natural-language instruction describing what to do. */
  instruction: string;
  /** For draft_reply: the body of the draft to create. */
  draftBody?: string;
}

/**
 * Structured outcome returned by the sub-agent after executing.
 */
export interface SubAgentOutcome {
  success: boolean;
  action_taken: string;
  messageId: string;
  error?: string;
  /** For draft_reply: the ID of the created draft. */
  draftId?: string;
}

/**
 * Result of a post-action verification step.
 */
export interface VerificationResult {
  verified: boolean;
  actual_state: string;
  messageId: string;
}

export interface GmailSubAgent {
  dispatch(instruction: SubAgentInstruction): Promise<SubAgentOutcome>;
  verify(messageId: string, expectedAction: string, meta?: { draftId?: string }): Promise<VerificationResult>;
}

/**
 * Create a Gmail sub-agent backed by the given Gmail provider.
 * The sub-agent accepts free-form instructions and returns structured outcomes.
 */
export function createGmailSubAgent(gmail: FakeGmail): GmailSubAgent {
  return {
    async dispatch(instruction: SubAgentInstruction): Promise<SubAgentOutcome> {
      if (instruction.capability === 'archive') {
        const found = gmail.archive(instruction.messageId);
        if (!found) {
          return {
            success: false,
            action_taken: 'archive',
            messageId: instruction.messageId,
            error: `message not found: ${instruction.messageId}`,
          };
        }
        return {
          success: true,
          action_taken: 'archive',
          messageId: instruction.messageId,
        };
      }

      if (instruction.capability === 'draft_reply') {
        const body = instruction.draftBody ?? instruction.instruction;
        const draft = gmail.createDraft(instruction.messageId, body);
        if (!draft) {
          return {
            success: false,
            action_taken: 'draft_reply',
            messageId: instruction.messageId,
            error: `message not found: ${instruction.messageId}`,
          };
        }
        return {
          success: true,
          action_taken: 'draft_reply',
          messageId: instruction.messageId,
          draftId: draft.id,
        };
      }

      return {
        success: false,
        action_taken: instruction.capability,
        messageId: instruction.messageId,
        error: `unknown capability: ${instruction.capability}`,
      };
    },

    async verify(messageId: string, expectedAction: string, meta?: { draftId?: string }): Promise<VerificationResult> {
      if (expectedAction === 'archive') {
        const msg = gmail.getById(messageId);
        if (!msg) {
          return { verified: false, actual_state: 'not_found', messageId };
        }
        return {
          verified: msg.archived === true,
          actual_state: msg.archived ? 'archived' : 'not_archived',
          messageId,
        };
      }

      if (expectedAction === 'draft_reply') {
        if (!meta?.draftId) {
          return { verified: false, actual_state: 'no_draft_id', messageId };
        }
        const draft = gmail.getDraft(meta.draftId);
        if (!draft) {
          return { verified: false, actual_state: 'draft_not_found', messageId };
        }
        return {
          verified: draft.inReplyTo === messageId,
          actual_state: 'draft_exists',
          messageId,
        };
      }

      return { verified: false, actual_state: 'unknown', messageId };
    },
  };
}

File: python/steward/gmail/subagent.py
"""Gmail sub-agent: dispatch + verify for archive, draft_reply, send_draft.

Takes any `GmailProvider` (FakeGmail for tests, RealGmail for production)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from steward.gmail.provider import GmailProvider


@dataclass
class GmailSubAgent:
    gmail: GmailProvider

    async def dispatch(self, instruction: dict[str, Any]) -> dict[str, Any]:
        cap = instruction.get("capability")
        message_id = instruction.get("messageId", "")

        if cap == "archive":
            found = self.gmail.archive(message_id)
            if not found:
                return {
                    "success": False,
                    "action_taken": "archive",
                    "messageId": message_id,
                    "error": f"message not found: {message_id}",
                }
            return {"success": True, "action_taken": "archive", "messageId": message_id}

        if cap == "draft_reply":
            body = instruction.get("draftBody") or instruction.get("instruction", "")
            draft = self.gmail.create_draft(message_id, body)
            if not draft:
                return {
                    "success": False,
                    "action_taken": "draft_reply",
                    "messageId": message_id,
                    "error": f"message not found: {message_id}",
                }
            return {
                "success": True,
                "action_taken": "draft_reply",
                "messageId": message_id,
                "draftId": draft["id"],
            }

        if cap == "send_draft":
            draft_id = instruction.get("draftId")
            if not draft_id:
                return {
                    "success": False,
                    "action_taken": "send_draft",
                    "messageId": message_id,
                    "error": "no draftId provided",
                }
            sent = self.gmail.send_draft(draft_id)
            if not sent:
                return {
                    "success": False,
                    "action_taken": "send_draft",
                    "messageId": message_id,
                    "error": f"draft not found or already sent: {draft_id}",
                }
            return {
                "success": True,
                "action_taken": "send_draft",
                "messageId": message_id,
                "draftId": sent["id"],
            }

        return {
            "success": False,
            "action_taken": cap or "unknown",
            "messageId": message_id,
            "error": f"unknown capability: {cap}",
        }

    async def verify(
        self,
        message_id: str,
        expected_action: str,
        meta: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        if expected_action == "archive":
            msg = self.gmail.get_by_id(message_id)
            if not msg:
                return {"verified": False, "actual_state": "not_found", "messageId": message_id}
            archived = msg.get("archived") is True
            return {
                "verified": archived,
                "actual_state": "archived" if archived else "not_archived",
                "messageId": message_id,
            }

        if expected_action == "draft_reply":
            draft_id = (meta or {}).get("draftId")
            if not draft_id:
                return {"verified": False, "actual_state": "no_draft_id", "messageId": message_id}
            draft = self.gmail.get_draft(draft_id)
            if not draft:
                return {"verified": False, "actual_state": "draft_not_found", "messageId": message_id}
            return {
                "verified": draft["inReplyTo"] == message_id,
                "actual_state": "draft_exists",
                "messageId": message_id,
            }

        if expected_action == "send_draft":
            draft_id = (meta or {}).get("draftId")
            if not draft_id:
                return {"verified": False, "actual_state": "no_draft_id", "messageId": message_id}
            draft = self.gmail.get_draft(draft_id)
            if not draft:
                return {"verified": False, "actual_state": "draft_not_found", "messageId": message_id}
            sent = draft.get("sent") is True
            return {
                "verified": sent,
                "actual_state": "sent" if sent else "not_sent",
                "messageId": message_id,
            }

        return {"verified": False, "actual_state": "unknown", "messageId": message_id}


def create_gmail_sub_agent(gmail: GmailProvider) -> GmailSubAgent:
    return GmailSubAgent(gmail=gmail)
